sandbox/host/src/sandbox/wasmi_backend.rs

```diff
@@ -18,393 +18,544 @@
+        log::trace!("Failed to call wasmi instantiate: {error:?}");
+        InstantiationError::Instantiation
+    })?;
+    let instance = instance_pre.start(&mut *store).map_err(|error| {
+        log::trace!("Failed to call wasmi start: {error:?}");
+        InstantiationError::StartTrapped
     })?;
 
-    Ok(sandbox_instance)
+    Ok(SandboxInstance {
+        backend_instance: BackendInstanceBundle::Wasmi {
+            instance,
+            store: context.store().clone(),
+        },
+    })
 }
 
-/// Invoke a function within a sandboxed module
-pub fn invoke(
-    instance: &SandboxInstance,
-    module: &sandbox_wasmi::ModuleRef,
-    export_name: &str,
-    args: &[Value],
-    supervisor_context: &mut dyn SupervisorContext,
-) -> std::result::Result<Option<Value>, error::Error> {
-    with_guest_externals(instance, |guest_externals| {
-        SupervisorContextStore::using(supervisor_context, || {
-            let args = args.iter().cloned().map(From::from).collect::<Vec<_>>();
-
-            module
-                .invoke_export(export_name, &args, guest_externals)
-                .map(|result| result.map(Into::into))
-                .map_err(|error| {
-                    if matches!(error, sandbox_wasmi::Error::Trap(Trap::Code(TrapCode::StackOverflow))) {
-                        // Panic stops process queue execution in that case.
-                        // This allows to avoid error lead to consensus failures, that must be handled
-                        // in node binaries forever. If this panic occur, then we must increase stack memory size,
-                        // or tune stack limit injection.
-                        // see also https://github.com/wasmerio/wasmer/issues/4181
-                        let err_msg = format!(
-                            "invoke: Suppose that this can not happen, because we have a stack limit instrumentation in programs. \
-                            Export name - {export_name}, args - {args:?}",
-                        );
-
-                        log::error!("{err_msg}");
-                        unreachable!("{err_msg}")
-                    }
-                    error::Error::Sandbox(error.to_string())
-                })
-        })
-    })
+fn dispatch_function(
+    supervisor_func_index: SupervisorFuncIndex,
+    store: &mut Store,
+    func_ty: &wasmi::FuncType,
+) -> wasmi::Func {
+    wasmi::Func::new(
+        store,
+        func_ty.clone(),
+        move |_caller, params, results| -> Result<(), wasmi::Error> {
+            SupervisorContextStore::with(|supervisor_context| {
+                let invoke_args_data = params
+                    .iter()
+                    .map(|value| {
+                        into_value(value).ok_or_else(|| {
+                            host_trap(format!("Unsupported function argument: {:?}", value))
+                        })
+                    })
+                    .collect::<std::result::Result<Vec<_>, _>>()?
```

Comment:
```suggestion
                    .collect::<Result<Vec<_>, _>>()?
```
This and other fully-qualified names can be omitted

Reply:
Fixed